subgraphs/arbitrum-one-bridge/src/common/utils.ts

```diff
@@ -0,0 +1,94 @@
+import { Token } from "../../generated/schema";
+import { bigIntToBigDecimal } from "../sdk/util/numbers";
+import { TokenPricer } from "../sdk/protocols/config";
+import { getUsdPrice, getUsdPricePerToken } from "../prices";
+import { TokenInitializer, TokenParams } from "../sdk/protocols/bridge/tokens";
+import { Address, BigDecimal, BigInt, ethereum } from "@graphprotocol/graph-ts";
+import { _ERC20 } from "../../generated/ERC20Gateway/_ERC20";
+import { BridgePermissionType } from "../sdk/protocols/bridge/enums";
+import { BridgeConfig } from "../sdk/protocols/bridge/config";
+import { Versions } from "../versions";
+import { ETH_ADDRESS } from "../sdk/util/constants";
+
+export class Pricer implements TokenPricer {
+  block: ethereum.Block;
+
+  constructor(block: ethereum.Block) {
+    this.block = block;
+  }
+
+  getTokenPrice(token: Token): BigDecimal {
+    const price = getUsdPricePerToken(Address.fromBytes(token.id), this.block);
+    return price.usdPrice;
+  }
+
+  getAmountValueUSD(token: Token, amount: BigInt): BigDecimal {
+    const _amount = bigIntToBigDecimal(amount, token.decimals);
+    return getUsdPrice(Address.fromBytes(token.id), _amount, this.block);
+  }
+}
+
+export class TokenInit implements TokenInitializer {
+  getTokenParams(address: Address): TokenParams {
+    const name = this.fetchTokenName(address);
+    const symbol = this.fetchTokenSymbol(address);
+    const decimals = this.fetchTokenDecimals(address) as i32;
+
+    return {
+      name,
+      symbol,
+      decimals,
+    };
+  }
+
+  fetchTokenName(tokenAddress: Address): string {
+    const tokenContract = _ERC20.bind(tokenAddress);
+    const call = tokenContract.try_name();
+    if (call.reverted) {
+      return tokenAddress.toHexString();
+    } else {
+      return call.value;
+    }
+  }
+
+  fetchTokenSymbol(tokenAddress: Address): string {
+    const tokenContract = _ERC20.bind(tokenAddress);
+    const call = tokenContract.try_symbol();
+    if (call.reverted) {
+      return " ";
+    } else {
+      return call.value;
+    }
+  }
+
+  fetchTokenDecimals(tokenAddress: Address): number {
+    const tokenContract = _ERC20.bind(tokenAddress);
+    const call = tokenContract.try_decimals();
+    if (call.reverted) {
+      return 0;
+    } else {
+      return call.value.toI32();
+    }
+  }
```

Comment:
Since you are also potentially dealing with native ETH, you'll want to handle that particular case.

ETH is not an ERC20, so to get the symbol, name, and decimals you cannot make a contract call. This is probably the reason why you have amounts and prices off by orders of magnitude.

Reply:
@jaimehgb  How do you recommend handling the pricing? Shouldn't native eth pricing be handled by `prices` lib given we can expect this to be needed in most mainnet based subgraphs?

I can think of two ways to manage pricing here:

1. Set token name (to ETH_ADDRESS), symbol (to ETH_SYMBOL), decimals (to 18) if the address is eth_address. Expect `prices` lib to return accurate price. Can `prices` lib return a price when the token info is set? Didn't find reference to code for `prices` lib handling this.

2. Update the `getTokenPrice` and `getAmountValueUSD` functions in the Pricer to get ETH price using WETH price if token address == ETH_ADDRESS (0xee...eee).

Reply:
I think what @jaimehgb is referring to is that we just need to handle the case when we get the eth address (generally it is (0xEeee... OR 0x0000...). In those cases we need to hardcode the id (0xEeee...eeee), name, symbol, and decimals. Without proper decimals it will make the prices off by A LOT.

As far as pricing goes, once the proper values for native ETH are hardcoded I'd say we can expect the price lib to work properly to get the eth price.

Reply:
Cool, that aligns with the approach 1 I outlined. Pushed this change to my target, will review and confirm.